### ISSUE TITLE
Naming Virtual Desktops

### ISSUE THREAD

**Issue by moser96** [Grabacr07/VirtualDesktop#17]
It would be neat if we could name or alias the virtual desktops.

**Comment by ericnewton76** [comment]
This is probably not supported by the underlying api, and therefore would not be implemented here... sorry!

**Comment by mzomparelli** [comment]
I don't even see the point of naming a desktop. The names are the numbers.

**Comment by MetaFight** [comment]
@moser96 I agree, naming them would be useful.  The Windows 10 Feedback app already has a few entries for this suggestion.  You should find them and upvote them.

**Comment by mzomparelli** [comment]
If anyone is interested, this app does useful things including naming desktops.

https://github.com/mzomparelli/zVirtualDesktop

**Comment by MetaFight** [comment]
@mzomparelli  Sorry, but your self-promotion feels a bit heavy handed and in poor taste.  Another project's issue tracker really isn't the place for ads.

**Comment by mzomparelli** [comment]
The app was built from this project and specifically does what this issue is asking. I don’t see a problem with my comment.

**Comment by Beej126** [comment]
> @ericnewton76: This is probably not supported by the underlying api, and therefore would not be implemented here... sorry!

[Windows 10 now supports this natively](https://docs.microsoft.com/en-us/windows-insider/at-home/whats-new-wip-at-home-20h1#update-on-virtual-desktop-renaming-build-18975)... will have to keep an eye out for if/where this gets exposed via apis

especially now that i've [created this taskbar widget](https://github.com/Beej126/Win10VirtualDesktopToolbar) for myself based on this lib... it would be nice to integrate with this new native Windows behavior... i.e. dump my names list and register for name change notifications from Windows

> @mzomparelli: I don't even see the point of naming a desktop. The names are the numbers.

seeing them titled at a glance helps me remember what the session is for ... hovering in the TaskView mode shows you the windows so you can get there pretty quick but the names are a nice little optimization... amazing msft applied the resources to make this happen out of all the priorities that fly

**Comment by lutz** [comment]
Currently with windows 2004 it is possible to get (not set)  the name for the current virtual desktop. I wrote about it in a other  [github project](https://github.com/MScholtes/VirtualDesktop/issues/10).

**Comment by Beej126** [comment]
@lutz, thanks for sharing! this is good news. i will incorporate into my [virtdesk taskbar widget project](https://github.com/Beej126/Win10VirtualDesktopDeskBand). please do include me if you find msft provides way to set the name at some point.

**Comment by lutz** [comment]
@Beej126 No problem currently it seems to be no way to set the name. You can manual edit the registry entry but this is not auto synced to the explorer (in my personal test)  I don\`t need the possibilty to set the name. I have a small [dev project](https://github.com/lutz/VirtualDesktopNameDeskband) which is  similar to your project. I want only a small name indicator. The rest of the functions (creating, switching and so on) i use the regular keybindings of windows (maybe i will add keybinding to `CTRL+WIN+[Number]` to switch to a specific desktop). If i have time to understand the code here for `VirtualDesktop.CurrentChanged` event i hope i can replace my global key hook.

**Comment by Beej126** [comment]
@lutz, that's awesome you independently picked up on the same general idea! deskband with name just feels so right since taskbar is the one static component during desktop changes... i couldn't resist implementing desktop change via dropdown in that space... it has quickly become instinctual which isn't always the case with new UI. i will build yours and check out UI =)

VirtualDesktop.CurrentChanged event is very straightforward ([implementation example](https://github.com/Beej126/Win10VirtualDesktopDeskBand/blob/b22010b2ae65cd1142edc593a9d04af5c3289cd7/Win10VirtualDesktopDeskBand/DeskBandUserControl.xaml.cs#L66))

**Comment by lutz** [comment]
> VirtualDesktop.CurrentChanged event is very straightforward (implementation example)

I meant the logic behind because i tied the dll from this project but ran into some issues like #44 and the event never raised when i comment the initialize methode out.

**Comment by MScholtes** [comment]
In Windows 10 20H1 there are two additional COM interfaces present that expand the original ones with functions to support desktop naming:

Interface **IVirtualDesktop2 (GUID {31EBDE3F-6EC3-4CBD-B9FB-0EF6D09B41F4})** with a new function you could declare as **void GetName([MarshalAs(UnmanagedType.HString)] out string name);**

Interface **IVirtualDesktopManagerInternal2 (GUID {0F3A72B0-4566-487E-9A33-4ED302F6D6CE})** with a new function you could declare as **void SetName(IVirtualDesktop desktop, [MarshalAs(UnmanagedType.HString)] string name);**

With those declarations desktop naming can be implemented without the need to recycle explorer (investigated with the great [OleViewDotNet](https://github.com/tyranid/oleviewdotnet)).

**Comment by lutz** [comment]
@MScholtes Big thanks for finding it out! This is a better way instead of look into the registry.

**Comment by Beej126** [comment]
this is killer! thank you for sharing @MScholtes !
